file/repair.go

```diff
@@ -24,51 +28,143 @@ func Repair(f *ast.File, importDir string, sn util.StructManager) bool {
 	}
 
-	ast.Inspect(f, func(n ast.Node) bool {
-		a, ok := n.(*ast.CompositeLit)
+	tokenFile := fset.File(f.Pos())
+	linePos := getLineTokenPositions(tokenFile)
+	parentStructure := map[ast.Node]ast.Node{}
+	offset := 0
+
+	astutil.Apply(f, func(c *astutil.Cursor) bool {
+		parentStructure[c.Node()] = c.Parent()
+		nudgeTokenPositions(c.Node(), int64(offset))
+
+		a, ok := c.Parent().(*ast.CompositeLit)
 		if !ok {
 			return true
 		}
 
-		for i, e := range a.Elts {
-			switch e.(type) {
-			case *ast.KeyValueExpr:
-			default:
-				once.Do(buildOutImports)
-
-				var importReference string
-				var structReference string
-
-				switch t := a.Type.(type) {
-				case *ast.SelectorExpr: // this struct declaration is import from another package
-					structName := t.Sel.String()
-
-					pkg, ok := t.X.(*ast.Ident)
-					if ok {
-						importReference = importsToPaths[util.RemoveQuotes(pkg.String())]
-						structReference = structName
-					}
-				case *ast.Ident: // this struct declaration is local to the package
-					importReference = importDir
-					structReference = t.Name
-				}
+		expr, ok := c.Node().(ast.Expr)
+		if !ok {
+			return true
+		}
+
+		switch expr.(type) {
+		case *ast.KeyValueExpr:
+		default:
+			once.Do(buildOutImports)
+
+			var importReference string
+			var structReference string
 
-				names, ok := sn.Get(importReference, structReference)
-				if !ok || len(names) <= i {
-					return true
+			switch t := a.Type.(type) {
+			case *ast.SelectorExpr: // this struct declaration is import from another package
+				structName := t.Sel.String()
+
+				pkg, ok := t.X.(*ast.Ident)
+				if ok {
+					importReference = importsToPaths[util.RemoveQuotes(pkg.String())]
+					structReference = structName
 				}
 
-				a.Elts[i] = &ast.KeyValueExpr{
-					Value: e,
-					Key: &ast.Ident{
-						Name: names[i],
-					},
-					Colon: f.End(),
+			case *ast.Ident: // this struct declaration is local to the package
+				importReference = importDir
+				structReference = t.Name
+			}
+
+			names, ok := sn.Get(importReference, structReference)
+			if !ok || len(names) <= c.Index() || c.Index() == -1 {
+				return true
+			}
+
+			nudge := len(names[c.Index()]) + 2
+			for i, l := range linePos {
+				if l > c.Node().Pos() {
+					linePos[i] = linePos[i] + token.Pos(nudge)
 				}
 			}
+
+			offset += nudge
+			nudgeDeclarationsAfter(f, nudge, expr.Pos())
+			nudgeTokenPositions(expr, int64(nudge))
+
+			k := &ast.KeyValueExpr{
+				Value: expr,
+				Key: &ast.Ident{
+					NamePos: expr.Pos() - 2 - token.Pos(len(names[c.Index()])),
+					Name:    names[c.Index()],
+				},
+				Colon: expr.Pos() - 2,
+			}
+			c.Replace(k)
+
+			for cur := c.Parent(); cur != nil; cur = parentStructure[cur] {
+				nudgeRightBrace(cur, int64(len(names[c.Index()])+2))
+			}
+		}
+
+		return true
+	}, nil)
+
+	rp := RepairInfo{
+		Lines: getLineOffsets(linePos, tokenFile),
+	}
+	return rp, nodesRepared
+}
+
+func nudgeTokenPositions(i interface{}, offset int64) {
```

Comment:
All of these nudges make me feel like there's a better way... But :shrug: if it works it works

Reply:
You and me both. It boils down to the fact that I am insert new content into the middle of an ast - and trying to respect the original line breaks and formatting. The token printers rely on the line offsets in the `token.File` for this file and the `base` and `size` as recorded in the `token.FileSet` in order to know on which lines which parts of the ast go.

If the printer finds an issue - either an invalid position or additional positions that do not make sense, it tries to  merge lines/insert lines at the wrong position. In master this results in all statements for a struct /usually/ ending up all on the same line. However, its anyone's guess as to what will actually happen. In a few test cases, I was able to get all by the last two brackets to end up on a the same line 🤷‍♂ . Most of these oddities are actually sorted out by the format pass through we do (so most of the issues don't present themselves in master).

Reply:
Also - this was better than the alternative - a BUNCH of type switches. And I mean A BUNCH. I eventually gave up on that. It was several hundred lines of type switches.